File: lib/doubly-linked-list/doubly_linked_list.rb
class DoublyLinkedList
  def initialize
    @head = nil
    @tail = nil
  end

  def insert_left(value)
    @head = Node.new(value, right: head)
    @tail ||= head

    head.right.left = head if head.right

    self
  end

  def insert_right(value)
    @tail = Node.new(value, left: tail)
    @head ||= tail

    tail.left.right = tail if tail.left

    self
  end

  def delete_left(value)
    node = each_left { |v| break v if v == value }

    delete_node(node)
  end

  def delete_right(value)
    node = each_right { |v| break v if v == value }

    delete_node(node)
  end

  def each_left
    node = head

    while node
      yield node.value
      node = node.right
    end
  end

  def each_right
    node = tail

    while node
      yield node.value
      node = node.left
    end
  end

  def inspect
    array = []
    each_left { |v| array << v }

    "<<#{array.join(", ")}>>"
  end

  private

  attr_reader :head, :tail

  def delete_node(node)
    return unless node

    if node == head && node == tail
      @head = @tail = nil
    elsif node == head
      @head = node.right
      head.left = nil
    elsif node == tail
      @tail = node.left
      tail.right = nil
    else
      node.left.right = node.right
      node.right.left = node.left
    end

    node.value
  end

  class Node
    attr_accessor :value, :left, :right

    def initialize(value, left: nil, right: nil)
      @value = value
      @left = left
      @right = right
    end
  end
end

File: lib/doubly-linked-list/doubly_linked_list.js
var DoublyLinkedList;

(function () {
  'use strict';

  DoublyLinkedList = function () {
    this.head = null;
    this.tail = null;
  };

  DoublyLinkedList.prototype.insertLeft = function (value) {
    var node = createNode(value);

    node.next = this.head;
    if (this.head) {
      this.head.previous = node;
    }
    this.head = node;
    if (this.tail === null) {
      this.tail = node;
    }

    return this;
  };

  DoublyLinkedList.prototype.insertRight = function (value) {
    var node = createNode(value);

    node.previous = this.tail;
    if (this.tail) {
      this.tail.next = node;
    }
    this.tail = node;
    if (this.head === null) {
      this.head = node;
    }

    return this;
  };

  DoublyLinkedList.prototype.eachLeft = function (callback) {
    eachNodeLeft(this.head, callback);
  };

  DoublyLinkedList.prototype.eachRight = function (callback) {
    eachNodeRight(this.tail, callback);
  };

  var createNode = function (value) {
    return { value: value, next: null, previous: null };
  };

  var eachNodeLeft = function (head, callback) {
    if (head === null) {
      return;
    }

    callback(head.value);
    eachNodeLeft(head.next, callback);
  };

  var eachNodeRight = function (tail, callback) {
    if (tail === null) {
      return;
    }

    callback(tail.value);
    eachNodeRight(tail.previous, callback);
  };
})();
